### ISSUE TITLE
Tree-ids

### ISSUE THREAD

**Issue by danpovey** [kaldi-asr/kaldi#2093]
I think we should have a concept of "tree-id", similar to ivector-id, that can be used to check compatibility of things like egs directories.  There are a bunch of errors that this could catch.  I don't know how to do it -based on md5sum?
Anyone interested?

**Comment by galv** [comment]
What errors would this catch? Is it normal for two separate trees to have the same num-pdfs, context-width and central-position, but still be different? I know it is for ivectors, since their training process isn't convex by virtual of the gaussian mixture.

I personally don't see why md5sum wouldn't work for creating the "tree.id" file, by the way.

**Comment by danpovey** [comment]
Yes it's definitely possible to have the same num-pdfs but still be
different; in general a rerun of the same recipe will give a different
tree, probably due to roundoff effects in stats accumulation.  This will be
even more likely after Vimal adds a change to make the num-pdfs a multiple
of 8 (this will be helpful in increasing GPU speed on Volta architectures).

**Comment by galv** [comment]
I'm curious about how this relates to NVIDIA's Volta architecture. Does it have to do with the alignment required for tensorcores? If it's interesting enough, I could be convinced to do this.

**Comment by danpovey** [comment]
Matrix multiplication only uses the tensorcores if the sizes involved are
multiples of 8 (or in one particular case, a multiple of 4).  So ensuring
the num-pdfs is a multiple of 8 will be helpful.  But I believe Vimal has
already worked out how to do this.

**Comment by galv** [comment]
I haven't looked into it since I haven't had access to a Volta GPU yet, but I'm a little skeptical that NVIDIA would require such requirements on sizes. After all, they could create 8x8 tiles of a larger matrix to then send through the specific tensorcore pathways. I'm a little disappointed if this isn't already hidden away by the CUBLAS API. But I am curious enough to take a look at how tensor cores work now.

**Comment by danpovey** [comment]
I seem to remember seeing that in the docs about the Volta architecture.
If I'm wrong about that I'd like to know.

**Comment by galv** [comment]
Okay, I was wrong. Take a look here:

> All of k, lda, ldb, and ldc must be a multiple of eight; m must be a multiple of four. The Tensor Core math routines stride through input data in steps of eight values, so the dimensions of the matrices must be multiples of eight.

[source](https://devblogs.nvidia.com/parallelforall/programming-tensor-cores-cuda-9/)

Note that this doesn't include `n`. I'm not sure what Vimal is doing, but if you can condition the problem such that the number of pdfs (==number of leaves in the tree, right?) corresponds to dimension n, then you may not have to worry about modifying the tree-creation code. I'm guessing that this has something to do with predicting a softmax over the pdfs. From what I remember, we often set m to be the batch size, and k to be the input feature size, and n to be the output size in linear layers, but I could be wrong. (I'm not sure if the fact that CUBLAS is column-major messes with this). I also haven't thought about the backwards pass yet.

@vimalmanohar Take a look above, and let me know if you want to discuss this further.

Also, feel free to enlist my help on anything related to CUDA or CUBLAS. I've improved a lot over the years, and this stuff interests me.

**Comment by danpovey** [comment]
I'm pretty sure once you take into account the natural gradient code,
there's going to be a matrix somewhere or other that doesn't satisfy those
properties unless the tree-dim is a multiple of 8.

This may actually be a good reason to keep the memory allocation code
as-is: I recently asked Hossein to check whether it was actually necessary
to use the cudaMallocPitch code (which is careful about ensuring the
strides are multiples of a high power of 2).  There didn't seem to be any
performance difference when using regular malloc.  But also, it didn't save
us memory.
It still seems to be the case that much less memory is available to be used
than the GPU claims-- less than half, if memory serves.  The problems are
usually when trying to allocate something quite large.  I don't know if
this is due to memory fragmentation or what.

**Comment by galv** [comment]
I'm not sure what natural gradient has to do with it, but I realize now that the output dimension of the softmax (n) will become m in the backpropagation step anyway, so I understand now that it is necessary to make the number of leaves in a tree a multiple of 8.

Memory allocation in basically all deep neural network libraries that I know of other than Kaldi use cudaMalloc everywhere. It prevents you from loading 4 words at a time from global memory, say at the beginning of a row, but this seems to be standard in deep neural network toolkits at this point.

cudaMallocPitch allows you to do things like load 4 words of global memory at a time from a row, but you need to provide a compiler hint (since it doesn't know whether a given pointer was cudaMalloc'd or cudaMallocPitch'd) via a reinterpret_cast<> (ugh...), so you're probably not getting all the benefit you would from it.

Less memory being available than the GPU claims is definitely odd. I don't know enough to comment further.
